src/compiler/checker.ts

```diff
@@ -3690,6 +3697,48 @@ module ts {
             }
             return links.resolvedType;
         }
+        
+        function getTypeGuardType(node: TypeGuardTypeNode): Type {
+            let type = <TypeGuardType>createType(TypeFlags.TypeGuard);
+            let hasParameters = false;
+            type.parameterIndex = -1;
+            if (node.target) {
+                let functionDeclaration = <SignatureDeclaration>node.target.parent.parent;
+                if (functionDeclaration.parameters) {
+                    for (let i = 0; i < functionDeclaration.parameters.length; i++) {
+                        let param = functionDeclaration.parameters[i];
+                        if (param.name.kind === SyntaxKind.Identifier && (<Identifier>param.name).text === node.target.text) {
+                            type.parameterIndex = i;
+                            break;
+                        }
+                    }
+                    hasParameters = true;
+                }
+                type.parameterName = node.target.text;
+            }
+            if (hasParameters && type.parameterIndex === -1) {

```

Comment:
what about the `this` case.. you need to check that it is used in a correct way, i.e. only within a class maybe..